core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -584,6 +612,45 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     }
   }
 
+  private def updateCatalogPermissions(catalogUpdates: Seq[WorkspaceCatalog], dataAccess: DataAccess, workspaceContext: SlickWorkspaceContext)= {
+    for {
+      refsToUpdateByEmail <- dataAccess.rawlsGroupQuery.loadRefsFromEmails(catalogUpdates.map(_.email))  // map from email to ref
+      (usersWithCatalog, groupsWithCatalog) <- dataAccess.workspaceQuery.findWorkspaceUsersAndGroupsWithCatalog(workspaceContext.workspaceId)  // refs with catalog
+    } yield {
+      val (emailsFound, emailsNotFound) = catalogUpdates.partition(catalogUpdate => refsToUpdateByEmail.keySet.contains(catalogUpdate.email))
+
+      val (emailsToAddCatalog, emailsToRemoveCatalog) = emailsFound.partition(updates => updates.catalog)
+
+      val usersToAdd = emailsToAddCatalog.collect {
+        case cat if refsToUpdateByEmail(cat.email).isLeft && !usersWithCatalog.contains(refsToUpdateByEmail(cat.email).left.get) =>
+          (refsToUpdateByEmail(cat.email).left.get, true)
+      }
+      val usersToRemove = emailsToRemoveCatalog.collect {
+        case cat if refsToUpdateByEmail(cat.email).isLeft && usersWithCatalog.contains(refsToUpdateByEmail(cat.email).left.get) =>
+          (refsToUpdateByEmail(cat.email).left.get, false)
+      }
+      val groupsToAdd = emailsToAddCatalog.collect {
+        case cat if refsToUpdateByEmail(cat.email).isRight && !groupsWithCatalog.contains(refsToUpdateByEmail(cat.email).right.get) =>
+          (refsToUpdateByEmail(cat.email).right.get, true)
+      }
+      val groupsToRemove = emailsToRemoveCatalog.collect {
+        case cat if refsToUpdateByEmail(cat.email).isRight && groupsWithCatalog.contains(refsToUpdateByEmail(cat.email).right.get) =>
+          (refsToUpdateByEmail(cat.email).right.get, false)
+      }
```

Comment:
bah. I spent a while trying to rewrite this to avoid all the boilerplate, and I didn't get that far. The best I could come up with was this:
```
      val (emailsFound, emailsNotFound) = catalogUpdates.partition(catalogUpdate => refsToUpdateByEmail.keySet.contains(catalogUpdate.email))

      val emailsFoundWithRefs:Seq[(WorkspaceCatalog,Either[RawlsUserRef, RawlsGroupRef])] = emailsFound map {wc =>
        (wc, refsToUpdateByEmail(wc.email))
      }

      val (emailsToAddCatalog, emailsToRemoveCatalog) = emailsFoundWithRefs.partition(update => update._1.catalog)

      val usersToAdd = emailsToAddCatalog.collect {
        case (cat, Left(userref)) if !usersWithCatalog.contains(userref) => (userref, true)
      }
      val usersToRemove = emailsToRemoveCatalog.collect {
        case (cat, Left(userref)) if usersWithCatalog.contains(userref) => (userref, false)
      }
      val groupsToAdd = emailsToAddCatalog.collect {
        case (cat, Right(groupref)) if !groupsWithCatalog.contains(groupref) => (groupref, true)
      }
      val groupsToRemove = emailsToRemoveCatalog.collect {
        case (cat, Right(groupref)) if groupsWithCatalog.contains(groupref) => (groupref, false)
      }
```

Your call on whether or not to implement this.